README.md
## 说明

- 插件会对`ip`和理由请求按照配置的规则进行判断，不符合的跳转到`ipRedirectUrl | requestRedirectUrl`，或者显示`ipCode, ipMessage`或者`requestCode, requestMessage`
- 如果`ip`和路由功能都开启，则先对`IP`进行判断，然后对路由请求进行判断，代码: `config.coreMiddleware.unshift(MIDDLEWARE_NAME_IP, MIDDLEWARE_NAME_REQUEST);`
- `ip`和路由如果有设置多个规则，则只要符合其中一个，就会结束判断
  - 规则 1: 60 秒内同一个 ip 访问 30 次，则禁止访问 300 秒
  - 规则 2: 80 秒内同一个 ip 访问 40 次，则禁止访问 400 秒
  - 在`ipRule`中，使用`Array.some`方法来依次判断，假设规则 1 符合，则禁止访问 300 秒，不会再判断规则 2
- 在您的应用中必须安装`egg-redis`，配置见下方

## 版本更新

- [CHANGELOG](./CHANGELOG.md)

## 依赖

> lodash, fast-json-stringify, uglify-es

## 安装

```bash
$ npm i egg-full-count egg-redis --save
```

## 使用

```js
// config/plugin.js
exports.fullFirewall = {
  enable: true,
  package: 'egg-full-count',
};

exports.redis = {
  enable: true,
  package: 'egg-redis',
};
```

## 配置

### 本插件配置 (必选)

```js
// {app_root}/config/config.default.js
exports.fullFirewall = {
  // 是否开启日志，日志位于 logs/你的项目名称/egg-web.log
  logEnable: false,
  // 是否检测IP
  useIP: true,
  // IP 检测规则, interval 时间间隔(秒), count 间隔内允许的次数, expire 封禁时间(秒)
  // 未配置 ipRule 则不会进行检测，但 ipDisabled 仍然有效
  ipRule: [
    // 60秒内请求超过3次就会被封禁300秒
    {
      interval: 60,
      count: 3,
      expire: 300,
    },
  ],
  // IP 被禁的时候，跳转的地址，优先级高于 ipCode和ipMessage，eg: '/ip-disabled'
  ipRedirectUrl: null,
  // IP 被禁的时候返回值
  ipCode: 403,
  // IP 被禁的时候返回的内容
  ipMessage: 'IP DISABLED',
  // 指定忽略检测的 IP列表
  // ipIgnore: ['127.0.0.1'],
  ipIgnore: [],
  // 直接封禁的 IP, useIP需要为 true
  ipDisabled: [],
  // IP 检测时忽略的路由, 支持以 * 结尾
  // 本功能主要用于 ipRedirectUrl, requestRedirectUrl
  // 如果这两个页面中需要请求资源，则会在 ip 判断中被阻止，ip 被阻止后会跳转到 ipRedirectUrl，造成死循环
  // 如果想控制这些资源，可以在 requestRule 进行配置
  ipIgnoreRequest: [
    {
      method: 'GET',
      url: '/build/*',
    },
    {
      method: '/favicon.ico',
      url: '/favicon.ico',
    },
  ],

  // 是否检测路由请求
  useRequest: true,
  // 路由请求 检测规则, urls 请求的路由集合, method 方法,interval 时间间隔(秒), count 间隔内允许的次数, expire 封禁时间(秒)
  requestRule: [
    // 示例: 60秒内用 GET 访问 / 或者 /404 3次以上会被封禁300秒
    {
      // 大写
      method: 'GET',
      interval: 60,
      count: 3,
      expire: 300,
      // 本规则应用到的路由
      urls: ['/', '/404'],
    },
  ],
  // 路由请求 被禁的时候，跳转的地址，优先级高于 requestCode 和 requestMessage ，eg: '/request-disabled'
  requestRedirectUrl: null,
  // 路由请求 被禁的时候返回值
  requestCode: 403,
  // 路由请求 被禁的时候返回的内容
  requestMessage: 'REQUEST DISABLED',
  // 忽略指定 IP 的路由请求
  // requestIgnoreIP: ['127.0.0.1'],
  requestIgnoreIP: [],

  // 本中间件使用 egg-redis，如果没有配置redis对应的名称，则默认使用 app.redis
  ipRedisName: null,
  requestName: null,
  // redis 键前缀
  ipRedisPrefix: 'ip:count',
  requestRedisPrefix: 'request:count',
};
```

### egg-redis 配置 1 (二选一)

```js
// 没有指定 ipRedisName 和 requestName, 则直接使用 app.redis
config.redis = {
  client: {
    port: 6379,
    host: '127.0.0.1',
    password: '123456',
    db: 0,
  },
};
```

### egg-redis 配置 2 (二选一)

```js
/**
 * 如果不希望本插件的缓存数据与应用的缓存混在一起，可以指定 ipRedisName 和 requestName
 * 比如指定 ipRedisName 为 firewallIP, requestName 为 firewallRequest
 * 不要忘记在 fullFirewall 配置中填上名称，ipRedisName和requestName默认都为null
 */

config.redis = {
  clients: {
    firewallIP: {
      port: 6379,
      host: '127.0.0.1',
      password: '123456',
      db: 0,
    },
    firewallRequest: {
      port: 6379,
      host: '127.0.0.1',
      password: '123456',
      db: 1,
    },
  },
};
```

## License

[MIT](LICENSE)


CHANGELOG.md
## 1.3.2

- 添加日志，配置新增`logEnable`，用于控制日志打印开关，默认关闭

## 1.3.3

- 修复`npm`发布时未包含`lib`文件夹的问题


test/full-firewall.test.js
'use strict';

const mock = require('egg-mock');

describe('test/full-firewall.test.js', () => {
  let app;
  before(() => {
    app = mock.app({
      baseDir: 'apps/full-firewall-test',
    });
    return app.ready();
  });

  after(() => app.close());
  afterEach(mock.restore);

  it('should GET /', () => {
    return app.httpRequest()
      .get('/')
      .expect('hi, fullFirewall')
      .expect(200);
  });
});


test/fixtures/apps/full-firewall-test/app/controller/home.js
'use strict';

const Controller = require('egg').Controller;

class HomeController extends Controller {
  async index() {
    this.ctx.body = 'hi, ' + this.app.plugins.fullFirewall.name;
  }
}

module.exports = HomeController;


app.js
'use strict';

const assert = require('assert');
const _ = require('lodash');

const MIDDLEWARE_NAME_IP = 'firewallIp';
const MIDDLEWARE_NAME_REQUEST = 'firewallRequest';

module.exports = app => {
    const {
        config,
    } = app;

    _.forEach([MIDDLEWARE_NAME_IP, MIDDLEWARE_NAME_REQUEST], name => {
        const index = config.coreMiddleware.indexOf(name);
        assert.equal(
            index,
            -1,
            `Duplication of middleware name found: ${name}. Rename your middleware other than "${name}" please.`
        );
    });

    // 配置检查与转换
    const {
        fullFirewall
    } = config;

    if (fullFirewall.useIP) {
        if (!_.isArray(fullFirewall.ipRule)) {
            fullFirewall.ipRule = [];
        }
        if (!_.isArray(fullFirewall.ipIgnore)) {
            fullFirewall.ipIgnore = [];
        }
        if (!_.isArray(fullFirewall.ipDisabled)) {
            fullFirewall.ipDisabled = [];
        }
        if (!_.isArray(fullFirewall.ipIgnoreRequest)) {
            fullFirewall.ipIgnoreRequest = [];
        } else {
            _.forEach(fullFirewall.ipIgnoreRequest, item => {
                if (_.isNil(item.method)) {
                    assert.fail('method(GET | POST | PUT ...) is required in every ipIgnoreRequest configuration');
                }
                item.method = _.toUpper(item.method);
            });
        }
        if (fullFirewall.ipRedirectUrl) {
            fullFirewall.ipIgnoreRequest.push({
                method: 'GET',
                url: fullFirewall.ipRedirectUrl,
            })
        }
    }

    if (fullFirewall.useRequest) {
        if (!_.isArray(fullFirewall.requestRule)) {
            fullFirewall.requestRule = [];
        } else {
            _.forEach(fullFirewall.requestRule, c => {
                if (_.isNil(c.method)) {
                    assert.fail('method(GET | POST | PUT ...) is required in every requestRule configuration');
                }
                c.method = _.toUpper(c.method);
            });
        }
        if (!_.isArray(fullFirewall.requestIgnoreIP)) {
            fullFirewall.requestIgnoreIP = [];
        }
    }

    // 尽量放置在最前面
    config.coreMiddleware.unshift(MIDDLEWARE_NAME_IP, MIDDLEWARE_NAME_REQUEST);
};

app/middleware/firewall-ip.js
'use strict';

const _ = require('lodash');
const fastJson = require('fast-json-stringify');
const Logger = require('../../lib/logger');

module.exports = (options, app) => {
  const {
    logEnable,
    useIP,
    ipRule,
    ipRedirectUrl,
    ipCode,
    ipMessage,
    ipIgnore,
    ipDisabled,
    ipRedisName,
    ipRedisPrefix,
    ipIgnoreRequest,
  } = app.config.fullFirewall;
  const redis = ipRedisName ? app.redis.get(ipRedisName) : app.redis;
  const logger = new Logger(app.coreLogger, logEnable, 'firewallIP');

  // 查找最大的 interval 做为缓存有效时间
  let cacheExpire = 0;
  _.forEach(ipRule, rule => {
    if (rule.interval > cacheExpire) {
      cacheExpire = rule.interval;
    }
  });

  // 预定义 JSON 转换结构
  const stringify = fastJson({
    title: 'Firewall IP Info',
    type: 'object',
    properties: {
      pret_t: {
        type: 'integer'
      },
      count: {
        type: 'integer'
      },
      disable_t: {
        type: 'integer'
      },
    }
  }, {
    uglify: true
  });

  const checkIP = async (ip, method, path) => {
    // 被禁止
    if (ipDisabled.some(val => val === ip)) {
      logger.log(`ip: ${ip} in the forbidden list`);
      return false;
    }
    // 未设置规则
    if (_.isEmpty(ipRule)) {
      return true;
    }
    // 忽略
    if (ipIgnore.some(val => val === ip)) {
      logger.log(`ip: ${ip} in the ignore list`);
      return true;
    }
    // 某些路由不受 IP 判断
    const ignoreRequestResult = ipIgnoreRequest.some(rule => {
      if (rule.method !== method) {
        return false;
      }
      const {
        url
      } = rule;
      if (url.charAt(url.length - 1) === '*') {
        const checkUrl = url.slice(0, url.length - 1);
        if (_.startsWith(path, checkUrl)) {
          return true;
        }
      } else if (url === path) {
        return true;
      }
      return false;
    })
    if (ignoreRequestResult) {
      logger.log(`path: ${path} in the ignore list`);
      return true;
    };

    const cacheKey = `${ipRedisPrefix}:${ip}`;
    const cacheValue = await redis.get(cacheKey);
    const info = JSON.parse(cacheValue);
    /**
     * info:
     * pre_t: 上一次检测的时间戳
     * count: 已访问的次数
     * disable_t: 封禁的时间戳
     */
    const t = _.parseInt(new Date() / 1000, 10);

    // 检测是否达到规则指定的条件
    if (info) {
      // 是否在封禁时间内
      if (info.disable_t > t) {
        logger.log(`ip: ${ip} is currently in the limit period`);
        return false;
      }

      info.count += 1;

      const result = ipRule.some(rule => {
        if (t - info.pret_t < rule.interval) {
          if (info.count > rule.count) {
            // 违反了本规则
            redis.set(cacheKey, stringify({
              pret_t: t,
              count: 0,
              disable_t: t + rule.expire,
            }), 'EX', cacheExpire);
            logger.log(`ip: ${ip} violation of rule ${JSON.stringify(rule)}`);
            return true;
          }
        }
        return false;
      });

      if (result) {
        // 违法了规则
        return false;
      }
      // 没有违反规则，则记录本次访问
      redis.set(cacheKey, stringify({
        pret_t: t,
        count: info.count,
        disable_t: 0,
      }), 'EX', cacheExpire);

    } else {
      // 没有任何记录，创建一个记录存储
      redis.set(cacheKey, stringify({
        pret_t: t,
        count: 1,
        disable_t: 0,
      }), 'EX', cacheExpire);
    }

    return true;
  };

  return async function firewallIP(ctx, next) {
    if (useIP) {
      const result = await checkIP(ctx.ip, ctx.request.method, ctx.path);
      if (!result) {
        if (ipRedirectUrl) {
          logger.log(`ip limit, redirect to ${ipRedirectUrl}`);
          await ctx.redirect(ipRedirectUrl);
        } else {
          ctx.status = ipCode;
          ctx.body = ipMessage;
          logger.log(`ip limit, status: ${ipCode}, body: ${JSON.stringify(ipMessage)}`);
        }
        return;
      }
    }
    await next();
  };
};

app/middleware/firewall-request.js
'use strict';

const _ = require('lodash');
const fastJson = require('fast-json-stringify');
const Logger = require('../../lib/logger');

module.exports = (options, app) => {
  const {
    logEnable,
    useRequest,
    requestRule,
    requestRedirectUrl,
    requestCode,
    requestMessage,
    requestIgnoreIP,
    requestName,
    requestRedisPrefix,
  } = app.config.fullFirewall;
  const redis = requestName ? app.redis.get(requestName) : app.redis;
  const logger = new Logger(app.coreLogger, logEnable, 'firewallRequest');

  // 查找最大的 interval 做为缓存有效时间
  let cacheExpire = 0;
  // 记录所有要检测的路由
  let urlList = [];
  _.forEach(requestRule, rule => {
    if (rule.interval > cacheExpire) {
      cacheExpire = rule.interval;
    }
    urlList = urlList.concat(rule.urls);
  });
  const urls = new Set(urlList);

  // 预定义 JSON 转换结构
  const stringify = fastJson({
    title: 'Firewall Request Info',
    type: 'object',
    properties: {
      pret_t: {
        type: 'integer'
      },
      count: {
        type: 'integer'
      },
      disable_t: {
        type: 'integer'
      },
    }
  }, {
    uglify: true
  });

  const checkRequest = async (ip, method, path) => {
    // 未设置规则
    if (_.isEmpty(requestRule)) {
      return true;
    }
    // 不在规则里面
    if (!urls.has(path)) {
      return true;
    }
    // 忽略
    if (requestIgnoreIP.some(val => val === ip)) {
      logger.log(`ip: ${ip} in the ignore list`);
      return true;
    }

    const cacheKey = `${requestRedisPrefix}:${ip}:${method}:${encodeURIComponent(path)}`;
    const cacheValue = await redis.get(cacheKey);
    const info = JSON.parse(cacheValue);
    /**
     * info:
     * pre_t: 上一次检测的时间戳
     * count: 已访问的次数
     * disable_t: 封禁的时间戳
     */
    const t = _.parseInt(new Date() / 1000, 10);

    // 检测是否达到规则指定的条件
    if (info) {
      // 是否在封禁时间内
      if (info.disable_t > t) {
        logger.log(`request: ${method} ${path} is currently in the limit period`);
        return false;
      }

      info.count += 1;

      const result = requestRule.some(rule => {
        if (rule.method === method) {
          if (rule.urls.some(url => url === path)) {
            // path 适用于本规则
            if (t - info.pret_t < rule.interval) {
              if (info.count > rule.count) {
                // 违反了本规则
                redis.set(cacheKey, stringify({
                  pret_t: t,
                  count: 0,
                  disable_t: t + rule.expire,
                }), 'EX', cacheExpire);
                logger.log(`request: ${method} ${path} violation of rule ${JSON.stringify(rule)}`);
                return true;
              }
            }
          }
        }
        return false;
      });

      if (result) {
        // 违法了规则
        return false;
      }
      // 没有违反规则，则记录本次访问
      redis.set(cacheKey, stringify({
        pret_t: t,
        count: info.count,
        disable_t: 0,
      }), 'EX', cacheExpire);

    } else {
      // 没有任何记录，创建一个记录存储
      redis.set(cacheKey, stringify({
        pret_t: t,
        count: 1,
        disable_t: 0,
      }), 'EX', cacheExpire);
    }

    return true;
  };

  return async function firewallRequest(ctx, next) {
    if (useRequest) {
      const result = await checkRequest(ctx.ip, ctx.request.method, ctx.path);
      if (!result) {
        if (requestRedirectUrl) {
          logger.log(`request limit, redirect to ${requestRedirectUrl}`);
          await ctx.redirect(requestRedirectUrl);
        } else {
          ctx.status = requestCode;
          ctx.body = requestMessage;
          logger.log(`request limit, status: ${ipCode}, body: ${JSON.stringify(ipMessage)}`);
        }
        return;
      }
    }
    await next();
  };
};

config/config.default.js
'use strict';

exports.fullFirewall = {
    // 是否开启日志，日志位于 logs/你的项目名称/egg-web.log
    logEnable: false,
    // 是否检测IP
    useIP: true,
    // IP 检测规则, interval 时间间隔(秒), count 间隔内允许的次数, expire 封禁时间(秒)
    // 未配置 ipRule 则不会进行检测，但 ipDisabled 仍然有效
    ipRule: [
        // 60秒内请求超过3次就会被封禁300秒
        {
            interval: 60,
            count: 3,
            expire: 300,
        },
    ],
    // IP 被禁的时候，跳转的地址，优先级高于 ipCode和ipMessage，eg: '/ip-disabled'
    ipRedirectUrl: null,
    // IP 被禁的时候返回值
    ipCode: 403,
    // IP 被禁的时候返回的内容
    ipMessage: 'IP DISABLED',
    // 指定忽略检测的 IP列表
    // ipIgnore: ['127.0.0.1'],
    ipIgnore: [],
    // 直接封禁的 IP, useIP需要为 true
    ipDisabled: [],
    // IP 检测时忽略的路由, 支持以 * 结尾
    // 本功能主要用于 ipRedirectUrl, requestRedirectUrl
    // 如果这两个页面中需要请求资源，则会在 ip 判断中被阻止，ip 被阻止后会跳转到 ipRedirectUrl，造成死循环
    // 如果想控制这些资源，可以在 requestRule 进行配置
    ipIgnoreRequest: [{
        method: 'GET',
        url: '/build/*',
    }, {
        method: 'GET',
        url: '/favicon.ico'
    }],

    // 是否检测路由请求
    useRequest: true,
    // 路由请求 检测规则, url(s) 请求的路由(们), method 方法,interval 时间间隔(秒), count 间隔内允许的次数, expire 封禁时间(秒)
    requestRule: [
        // 示例: 60秒内用 GET 访问 / 或者 /404 3次以上会被封禁300秒
        {
            // 大写
            method: 'GET',
            interval: 60,
            count: 3,
            expire: 300,
            // 本规则应用到的路由
            urls: ['/', '/404'],
        },
    ],
    // 路由请求 被禁的时候，跳转的地址，优先级高于 requestCode 和 requestMessage ，eg: '/request-disabled'
    requestRedirectUrl: null,
    // 路由请求 被禁的时候返回值
    requestCode: 403,
    // 路由请求 被禁的时候返回的内容
    requestMessage: 'REQUEST DISABLED',
    // 忽略指定 IP 的路由请求
    // requestIgnoreIP: ['127.0.0.1'],
    requestIgnoreIP: [],

    // 本中间件使用 egg-redis，如果没有配置redis对应的名称，则默认使用 app.redis
    ipRedisName: null,
    requestName: null,
    // redis 键前缀
    ipRedisPrefix: 'ip:count',
    requestRedisPrefix: 'request:count',
};